src/main/java/au/org/ala/biocache/dao/SearchDAOImpl.java

```diff
@@ -911,120 +1096,118 @@ public void writeCoordinatesToStream(SearchRequestParams searchParams,OutputStre
+                                        // throttle the download by sleeping
+                                        Thread.sleep(localThrottle);
                                     }
                                     qr = runSolrQuery(splitByFacetQuery, null, downloadBatchSize, startIndex, "_docid_", "asc");
                                 } else {
                                     qr = null;
                                 }
                             }
-                            return recordsForThread;
+                            return recordsForThread.get();
                         }
                     };
-                    futures.add(pool.submit(solrCallable));
+                    solrCallables.add(solrCallable);
                 }
     
-                //check the futures until all have finished
-                int totalDownload = 0;
-                Set<Future<Integer>> completeFutures = new HashSet<Future<Integer>>();
-                boolean allComplete = false;
-                while(!allComplete){
-                    for(Future future: futures){
-                        if(!completeFutures.contains(future)){
-                            if(future.isDone()){
-                                totalDownload += (Integer) future.get();
-                                completeFutures.add(future);
-                            }
-                        }
-                    }
-                    allComplete = completeFutures.size() == futures.size();
-                    if(!allComplete){
-                        Thread.sleep(1000);
+                // TODO: Do we support downloads running for longer than 1 day??
+                List<Future<Integer>> futures = nextExecutor.invokeAll(solrCallables, 1, TimeUnit.DAYS);
+                AtomicInteger totalDownload = new AtomicInteger(0);
+                for(Future<Integer> future: futures){
+                    if(future.isDone()){
+                        totalDownload.addAndGet(future.get());
                     }
                 }
-                pool.shutdown();
-                out.flush();
 
                 long finish = System.currentTimeMillis();
                 long timeTakenInSecs = (finish-start)/1000;
                 if(timeTakenInSecs <= 0) timeTakenInSecs = 1;
-                logger.info("Download of " + resultsCount + " records in " + timeTakenInSecs + " seconds. Record/sec: " + resultsCount.intValue()/timeTakenInSecs);
+                if(logger.isInfoEnabled()) {
+                    logger.info("Download of " + resultsCount + " records in " + timeTakenInSecs + " seconds. Record/sec: " + resultsCount.intValue()/timeTakenInSecs);
+                }
             } finally {
-                rw.finalise();
+                try {
+                    concurrentWrapper.finalise();
+                } finally {
+                    try {
+                        writerThread.join(writerFinalisationWaitMillis);
```

Comment:
In the normal case, I am going with an AtomicBoolean that is set to true at the end of the finalise method as the normal completion. Also wrapping the finalise method in another AtomicBoolean with compareAndSet(false, true) to ensure the finalisation is safe to call multiple times without corrupting the file.

In the slow case, I am going with a new configuration property downloadMaxCompletionTime, by default 5 minutes, which should be enough in normal circumstances to write the last 1000 items off the queue, since no new items are being added to the queue at this stage, but it is all configurable and we can change the defaults to match experience once it is tested.

In the slow case, any and all interrupts, or the writerThread completing for any reason, will also finish the section and trigger a "shutdownNow"/"hard shutdown" attempt just after this to further cleanup if we made it out of this stage while still holding resources.